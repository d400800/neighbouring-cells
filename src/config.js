export const colorMap = new Map([
    ["teal", "#19FA8C"],
    ["blue", "#4C32FA"],
    ["yellow", "#FADA0C"],
    ["red", "#FA1E00"],
]);

export const colors = Array.from(colorMap.keys());