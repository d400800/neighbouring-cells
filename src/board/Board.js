import {colors} from '../config.js';
import Utils from '../game/utils';

class Board {
    constructor(options) {
        this.defaultCellColors = options.colors;

        this.previousCellsToRepaint = [];
    }

    generateRandomBoard(size, colorMap) {
        const cellColors = Array.from(colorMap.keys());

        let board = [];

        for (let i = 0; i < size; ++i) {
            let row = [];

            for (let j = 0; j < size; ++j) {
                row.push(Utils.getRandomItem(cellColors || this.defaultCellColors));
            }

            board.push(row);
        }

        return board;
    }

    // bootleg bucket fill algorithm, don't judge
    repaint(color, originalBoard) {
        const board = originalBoard.slice();

        const visitedCells = new Set(['0|0']);

        const initialColor = board[0][0];

        let currentCell = [0,0];

        const findCellsToRepaint = (startCell) => {
            let neighbours = this.getNeighbours(board.length-1, board, initialColor, startCell, visitedCells);

            if (neighbours.length > 0) {
                for (const cell of neighbours) {
                    visitedCells.add(this.cellToString(cell));

                    findCellsToRepaint(cell);
                }
            }

            return visitedCells;
        };

        console.time('findCellsToRepaint');

        const cellsToRepaint = findCellsToRepaint(currentCell);

        //console.log('previousCellsToRepaint', this.previousCellsToRepaint);
        //this.previousCellsToRepaint = cellsToRepaint;

        console.timeEnd('findCellsToRepaint');

        for (const cell of Array.from(cellsToRepaint.values())) {
            const [row, col] = cell.split('|');

            board[row][col] = color;
        }

        return board;
    }

    cellToString([row, col]) {
        return `${row}|${col}`;
    }

    getNeighbours(size, board, color, [row, col], visitedCells) {
        const right = row < size ? [row+1, col] : [];
        const left = row !== 0 ? [row-1, col] : [];
        const top = col !== 0 ? [row, col-1] : [];
        const bottom = col < size ? [row, col+1] : [];

        return [right, left, top, bottom]
            .filter(cell => cell.length > 0 &&
                board[cell[0]][cell[1]] === color &&
                !visitedCells.has(this.cellToString(cell))
            );
    }

    isHomogeneous(board) {
        const initialCell = board[0][0];

        return board
            .filter(row => row.filter(cell => !(cell === initialCell)).length > 0)
            .length === 0;
    }
}

const board = new Board({colors});

export default board;