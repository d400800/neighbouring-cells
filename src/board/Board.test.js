import {defaultColorMap} from "../config";
import board from "./Board";

const myBoard = board.generateRandomBoard(18, defaultColorMap);

describe('testing the Board', () => {
    it('should create a Board', () => {
        const newBoard = board.repaint('teal', myBoard);

        expect(newBoard).toBeDefined();
    });
});