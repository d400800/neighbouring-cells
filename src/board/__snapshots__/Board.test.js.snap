// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the Board should verify that the board behaves consistently 1`] = `
Array [
  Array [
    "teal",
    "teal",
    "teal",
    "teal",
    "blue",
    "red",
    "red",
    "yellow",
    "teal",
    "blue",
    "teal",
    "yellow",
    "blue",
    "teal",
    "teal",
    "yellow",
    "teal",
    "blue",
  ],
  Array [
    "teal",
    "teal",
    "teal",
    "teal",
    "blue",
    "blue",
    "teal",
    "teal",
    "red",
    "teal",
    "teal",
    "red",
    "red",
    "yellow",
    "red",
    "yellow",
    "red",
    "blue",
  ],
  Array [
    "teal",
    "red",
    "teal",
    "blue",
    "blue",
    "yellow",
    "red",
    "red",
    "red",
    "yellow",
    "teal",
    "teal",
    "yellow",
    "yellow",
    "yellow",
    "red",
    "yellow",
    "teal",
  ],
  Array [
    "teal",
    "teal",
    "red",
    "red",
    "red",
    "teal",
    "teal",
    "red",
    "red",
    "blue",
    "yellow",
    "red",
    "red",
    "teal",
    "blue",
    "blue",
    "blue",
    "red",
  ],
  Array [
    "teal",
    "blue",
    "teal",
    "blue",
    "red",
    "teal",
    "teal",
    "blue",
    "blue",
    "yellow",
    "yellow",
    "red",
    "red",
    "red",
    "teal",
    "teal",
    "red",
    "red",
  ],
  Array [
    "teal",
    "teal",
    "red",
    "blue",
    "teal",
    "blue",
    "yellow",
    "yellow",
    "blue",
    "teal",
    "blue",
    "teal",
    "teal",
    "blue",
    "teal",
    "teal",
    "teal",
    "teal",
  ],
  Array [
    "yellow",
    "red",
    "red",
    "yellow",
    "yellow",
    "teal",
    "yellow",
    "yellow",
    "blue",
    "yellow",
    "yellow",
    "teal",
    "blue",
    "teal",
    "teal",
    "red",
    "teal",
    "teal",
  ],
  Array [
    "blue",
    "yellow",
    "red",
    "red",
    "yellow",
    "teal",
    "yellow",
    "blue",
    "teal",
    "yellow",
    "yellow",
    "red",
    "blue",
    "yellow",
    "red",
    "teal",
    "blue",
    "teal",
  ],
  Array [
    "red",
    "yellow",
    "teal",
    "blue",
    "teal",
    "blue",
    "blue",
    "teal",
    "yellow",
    "teal",
    "blue",
    "red",
    "yellow",
    "blue",
    "blue",
    "teal",
    "teal",
    "teal",
  ],
  Array [
    "red",
    "yellow",
    "red",
    "teal",
    "yellow",
    "blue",
    "yellow",
    "red",
    "blue",
    "teal",
    "yellow",
    "red",
    "red",
    "teal",
    "blue",
    "teal",
    "yellow",
    "teal",
  ],
  Array [
    "yellow",
    "blue",
    "blue",
    "red",
    "teal",
    "teal",
    "red",
    "teal",
    "teal",
    "teal",
    "blue",
    "blue",
    "blue",
    "teal",
    "red",
    "blue",
    "blue",
    "blue",
  ],
  Array [
    "blue",
    "teal",
    "blue",
    "red",
    "blue",
    "red",
    "red",
    "blue",
    "blue",
    "blue",
    "teal",
    "yellow",
    "blue",
    "yellow",
    "blue",
    "teal",
    "teal",
    "yellow",
  ],
  Array [
    "blue",
    "teal",
    "blue",
    "red",
    "blue",
    "teal",
    "blue",
    "teal",
    "teal",
    "blue",
    "teal",
    "blue",
    "blue",
    "teal",
    "teal",
    "blue",
    "red",
    "blue",
  ],
  Array [
    "blue",
    "blue",
    "teal",
    "teal",
    "blue",
    "red",
    "red",
    "yellow",
    "red",
    "red",
    "blue",
    "red",
    "red",
    "blue",
    "yellow",
    "red",
    "teal",
    "teal",
  ],
  Array [
    "blue",
    "red",
    "yellow",
    "blue",
    "teal",
    "teal",
    "teal",
    "yellow",
    "red",
    "blue",
    "teal",
    "teal",
    "teal",
    "red",
    "yellow",
    "blue",
    "teal",
    "teal",
  ],
  Array [
    "blue",
    "red",
    "red",
    "yellow",
    "yellow",
    "yellow",
    "red",
    "teal",
    "teal",
    "teal",
    "teal",
    "blue",
    "red",
    "blue",
    "red",
    "yellow",
    "teal",
    "red",
  ],
  Array [
    "teal",
    "yellow",
    "yellow",
    "yellow",
    "teal",
    "teal",
    "teal",
    "teal",
    "blue",
    "red",
    "yellow",
    "blue",
    "teal",
    "blue",
    "teal",
    "blue",
    "yellow",
    "teal",
  ],
  Array [
    "blue",
    "yellow",
    "teal",
    "yellow",
    "yellow",
    "teal",
    "yellow",
    "teal",
    "blue",
    "blue",
    "yellow",
    "blue",
    "blue",
    "yellow",
    "teal",
    "blue",
    "yellow",
    "red",
  ],
]
`;
